### ISSUE TITLE
Excluding files from workspace seems to be ignored

### ISSUE THREAD

**Issue by dsyzling** [microsoft/pyright#2108]
I'm trying to exclude files within certain directories from being included within the workspace. Specifically I have a .pytest_cache/ directory at the root of the project that triggers the following message whenever I run tests:

```
[Trace - 11:17:10 am] Sending notification 'workspace/didChangeWatchedFiles'.
Params: {
  "changes": [
    {
      "type": 2,
      "uri": "file:///home/dsyzling/projects/python/language/.pytest_cache/v/cache/nodeids"
    }
  ]
}
```

I add a pyrightconfig.json file to the root of my project with the following lines:

```json
{
    "exclude": [
        "**/.pytest_cache"
    ]
}
```
But even running pyright from the command line with pyright --verbose -w  I can see it triggering a change to files within this directory. 

Initially I discovered this because I wanted to optionally run mypy, however that equally creates directories and triggers file change notifications:

```
[Trace - 09:19:09 am] Sending notification 'workspace/didChangeWatchedFiles'.
Params: {
  "changes": [
    {
      "type": 1,
      "uri": "file:///home/dsyzling/projects/python/language/tests/.mypy_cache/3.8/%40plugins_snapshot.json"
    }
  ]
}
```

This trigger then causes a publish diagnostics message to be generated and Emacs/flycheck enters a loop of running diags with mypy as well as via the lsp server. I've tried disabling the .mypy_cache directory in a similar way to the above with a json and toml files without success.

I'm clearly doing something obviously wrong or missing something obvious. 

I understand there's a duplication of checks here, however I might want to run mypy type checking instead of pyright, and use pyright for the remaining lsp server functionality. 
 
pyright version is 1.1.156
python 3.8.3 (in a conda virtual environment)
Running on Arch Linux
I'm using Emacs, lsp-mode.

**Comment by erictraut** [comment]
The exclude list (along with the include list) tells pyright which source files to consider "part of the workspace". If an excluded file is imported by an included file, it will still need to be opened and partially analyzed to determine dependent type information when analyzing an included file. And if such a file is modified, files that depend upon it will need to be reanalyzed and a new diagnostics list published.

The exclude list does not mean "never open source files in this list". It also doesn't mean "don't install file watches for files or directories in this list".

So I think pyright is doing the right thing here.

You said:

> Emacs/flycheck enters a loop of running diags with mypy as well as via the lsp server

That sounds like a bug in emacs or in the mechanism that invokes mypy from emacs. Why is mypy getting invoked in response to diagnostics published from pyright?

**Comment by dsyzling** [comment]
Thanks for the reply, I thought the exclude list could be used to filter parts of your workspace that you didn't want to be part of the lsp server or checker process and were strictly not part of your 'project'. In this case json files are being generated and pyright is sending a diagnostics published response because it perceives they are source files that the workspace/project is interested in.

I agree that one approach to resolving this would be to prevent the Emacs side from responding to this notification. However it's more complicated because we have two approaches to linting errors - lsp diagnostics which are asynchronous and a chain of custom checkers that the user can configure - i.e. pylint, mypy, flake8. You either explicitly request linting checks to occur or they are hooked to an event when checks need to take place - in this case when an 'interesting' file has been modified and checks may need to run - workspace change. It's made more complicated because these two parts are developed independently, and made to cooperate but they each also have their own individual use cases and users. They are executed asynchronously and the results combined - since they may all reply at different times or may never reply. The results are displayed as if all the messages came from the lsp server. The old pyls used to be straightforward, you told that server to enable flake8, pylint etc. and it returned all of the warnings in an event or sequence of events. When the server doesn't support linters internally - the Emacs lsp side needs to start those as separate processes and collate the results. 

Is a publish diagnostic message generated if a file configured as being a 'source' file is changed - i.e. what stops a .pyc file being updated and sending a notification vs a json file? Presumably there's a list of file extensions that are regarded as project source files? 

No criticism but does it not seem redundant to generate a workspace/didChangeWatchedFiles for files that might be generated by tools such as test runners, linters  - just because they are being generated into hidden folders within the workspace tree? I appreciate some of these tools can be configured to generate their files elsewhere but it does maintain some form of scope for the project. 

Something else that occurred to me to research is whether there is a custom client workspace ignore list within Emacs lsp-mode - if I could add the directories to that list it might filter workspace change and publish diagnostic messages. 

Please don't take this as me campaigning for a functional change here, I'm just seeking to understand the current behaviour implemented.

**Comment by erictraut** [comment]
VS Code performs a similar "multiplexing" of language servers. It too invokes them asynchronously and combines their outputs, so the mechanism you're describing is familiar to me.

It makes sense that the chain of custom checkers would run when a "file changed" notification is received from the file system indicating that a relevant file has been modified, added or deleted. It doesn't make sense to me that these custom checkers would be triggered to run in response to diagnostics from another language server like pyright. Am I correct in understanding that's what happens here? If so, why is this being done? Any language server is allowed to publish a new diagnostic list whenever it sees fit to do so. It shouldn't have any impact on other language servers or linters. If it does, that sounds like a significant design flaw.

> ...does it not seem redundant to generate a workspace/didChangeWatchedFiles for files that might be generated by tools...

Pyright needs to monitor many different types of files and directories that can affect type analysis. It also needs to work with multiple different file change notification mechanisms because the LSP interface supports "workspace/didChangeWatchedFiles" messages for files and directories within the root directory but not outside that directory, plus pyright needs to run in "stand-alone" mode as a CLI. For these reasons, it installs a general file system watcher hook and filters the responses. So yes, it might end up ignoring some of the file change notifications that it receives. That's by design.

**Comment by dsyzling** [comment]
I think this is my misunderstanding of the process and configuration of client and server responsibilities. Emacs lsp-mode has an option to register file watchers for files in the workspace and I've now seen this as a capability which is negotiated as part of initialisation according to the lsp spec. The option is enabled by default and negotiated during initialisation, so the client is performing file watching. I think what was happening was that the client was generating the workspace changes as a result of the mypy (and pytest) files being created/changed which was then triggering a cascade of events leading to diagnostics being run again, causing looping.

Apologies for the confusion - I've simply added the pytest and mypy cache directories to this list and the validation chain now works without looping. I may discuss with the authors whether any of these common directories need to be added as part of some of the client side packages automatically.
